Add Seccomp Notify support

alban opened the pull request:
This adds the specification for Seccomp Userspace Notification and the Golang bindings. This contains:
- A new OCI hook "sendSeccompFd" used to pass the seccompfd to an external seccomp agent via the hook.
- Additional SeccompState struct containing the container state and file descriptors passed for seccomp.

This was discussed in the OCI Weekly Discussion on September 16th, 2020, see:
- https://hackmd.io/El8Dd2xrTlCaCG59ns5cwg#September-16-2020
- https://docs.google.com/document/d/1xHw5GQjMj6ZKR-40aKmTWZRkvlPuzMGQRu-YpOFQc30/edit

Documentation for this feature:
- https://www.kernel.org/doc/html/v5.0/userspace-api/seccomp_filter.html#userspace-notification
- man pages: seccomp_user_notif.2 at https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/log/?h=seccomp_user_notif
- brauner's blog: https://brauner.github.io/2020/07/23/seccomp-notify.html

This PR is an alternative proposal to PR 1038.

Signed-off-by: Alban Crequy <alban@kinvolk.io>

alban committed:
Add Seccomp Notify support

This adds the specification for Seccomp Userspace Notification and the
Golang bindings. This contains:
- A new OCI hook "sendSeccompFd" used to pass the seccompfd to an
  external seccomp agent via the hook.
- Additional SeccompState struct containing the container state and file
  descriptors passed for seccomp.

This was discussed in the OCI Weekly Discussion on September 16th, 2020, see:
- https://hackmd.io/El8Dd2xrTlCaCG59ns5cwg#September-16-2020
- https://docs.google.com/document/d/1xHw5GQjMj6ZKR-40aKmTWZRkvlPuzMGQRu-YpOFQc30/edit

Documentation for this feature:
- https://www.kernel.org/doc/html/v5.0/userspace-api/seccomp_filter.html#userspace-notification
- man pages: seccomp_user_notif.2 at https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/log/?h=seccomp_user_notif
- brauner's blog: https://brauner.github.io/2020/07/23/seccomp-notify.html

This PR is an alternative proposal to PR 1038.

Signed-off-by: Alban Crequy <alban@kinvolk.io>

mauriciovasquezbernal commented on config.md:
```diff
@@ -476,6 +481,16 @@ Cleanup or debugging functions are examples of such a hook.
 
+### <a name="configHooksSendSeccompFd" />SendSeccompFd
+
+The `sendSeccompFd` hooks is only called if the seccomp policy contains `SCMP_ACT_NOTIFY`.
```
```suggestion
The `sendSeccompFd` hooks MUST be only called if the seccomp policy contains `SCMP_ACT_NOTIFY`.
```

mauriciovasquezbernal commented on config.md:
```diff
@@ -476,6 +481,16 @@ Cleanup or debugging functions are examples of such a hook.
+The goal of this hook is to pass the seccomp file descriptor to a seccomp agent.
+
+The `sendSeccompFd` hooks' path MUST resolve in the [runtime namespace](glossary.md#runtime-namespace).
+The `peccompFdoststop` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
```
```suggestion
The `sendSeccompFd` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
```

:laughing:

mauriciovasquezbernal commented on config.md:
```diff
@@ -476,6 +481,16 @@ Cleanup or debugging functions are examples of such a hook.
 The `poststop` hooks' path MUST resolve in the [runtime namespace](glossary.md#runtime-namespace).
 The `poststop` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
 
+### <a name="configHooksSendSeccompFd" />SendSeccompFd
```
What about adding it before the `StartContainer Hooks` to keep them in the same order they are called?

mauriciovasquezbernal commented on config.md:
```diff
@@ -488,6 +503,7 @@ See the below table for a summary of hooks and when they are called:
 | `startContainer`        | container | After the start operation is called but before the user-specified program command is executed.                                     |
 | `poststart`             | runtime   | After the user-specified process is executed but before the start operation returns.                                               |
 | `poststop`              | runtime   | After the container is deleted but before the delete operation returns.                                                            |
+| `sendSeccompFd`         | runtime   | After the start operation is called but before the user-specified program command is executed.                                     |
```
Same here, what about putting this in the same order they are called?

mauriciovasquezbernal commented on specs-go/state.go:
```diff
@@ -32,4 +32,6 @@ type State struct {
 	Bundle string `json:"bundle"`
 	// Annotations are key values associated with the container.
 	Annotations map[string]string `json:"annotations,omitempty"`
+
```
```suggestion
  SeccompFd is the file descriptor of the seccomp filter associated to the container.
```

mauriciovasquezbernal commented on runtime.md:
```diff
@@ -29,6 +29,7 @@ The state of a container includes the following properties:
     This is provided so that consumers can find the container's configuration and root filesystem on the host.
 * **`annotations`** (map, OPTIONAL) contains the list of annotations associated with the container.
     If no annotations were provided then this property MAY either be absent or an empty map.
+* **`seccomp`** (map, OPTIONAL) contains additional TODO.
```
```suggestion
* **`seccompFd`** (int, OPTIONAL) contains the file descriptor of the seccomp filter associated to the container when it has any `SCMP_ACT_NOTIFY` action. 
```

Perhaps we should also say it's only relevant when calling the sendSeccompFd hook?

mauriciovasquezbernal commented on config-linux.md:
```diff
@@ -642,6 +642,7 @@ The following parameters can be specified to set up seccomp:
         * `SCMP_ACT_TRACE`
         * `SCMP_ACT_ALLOW`
         * `SCMP_ACT_LOG`
+        * `SCMP_ACT_NOTIFY`
```
Update "libseccomp v2.4.0" -> "libseccomp v2.5.0" above too.

mauriciovasquezbernal reviewed (commented):
Some minor comments and suggestions.

alban commented:
Thanks for the review @mauriciovasquezbernal !

I addressed your comments. Another change I'd like to make is on the State, to make it look like in the [proposal](https://docs.google.com/document/d/1xHw5GQjMj6ZKR-40aKmTWZRkvlPuzMGQRu-YpOFQc30/edit#):

```
    “seccomp”: {
        "fds": {
            seccomp-fd: 3
        },
        pid: 4499    # this is the process on which the seccomp policy is attached to (!=pid1)
}
```

~Also, the Travis tests failed, but I am not sure why. I'll check that.~
Edit: Travis tests to be fixed in upstream PR 1072.

alban committed:
travis: fix go_import_path

Without this, Travis CI only works on the main repository
(github.com/opencontainers/runtime-spec) but not on forks. It is useful
to make Travis CI work on forks for contributors to be able to test
their work before submitting a PR upstream.

See:
https://docs.travis-ci.com/user/languages/go#go-import-path

Symptoms:
```
$ make docs
go run ./.tool/version-doc.go > version.md
.tool/version-doc.go:10:2: cannot find package "github.com/opencontainers/runtime-spec/specs-go" in any of:
	/home/travis/.gimme/versions/go1.11.13.linux.amd64/src/github.com/opencontainers/runtime-spec/specs-go (from $GOROOT)
	/home/travis/gopath/src/github.com/opencontainers/runtime-spec/specs-go (from $GOPATH)
Makefile:53: recipe for target 'version.md' failed
make: *** [version.md] Error 1
The command "make docs" exited with 2.
```

Signed-off-by: Alban Crequy <alban@kinvolk.io>

mauriciovasquezbernal commented on config.md:
```diff
@@ -415,7 +420,8 @@ For POSIX platforms, the configuration structure supports `hooks` for configurin
 Hooks MUST be called in the listed order.
-The [state](runtime.md#state) of the container MUST be passed to hooks over stdin so that they may do work appropriate to the current state of the container.
+All hooks MUST be passed a data structure over stdin so that they may do work appropriately.
+Exect when specified otherwise above, the data structure is the [state](runtime.md#state) of the container.
```
```suggestion
Except when specified otherwise above, the data structure is the [state](runtime.md#state) of the container.
```

mauriciovasquezbernal commented on config.md:
```diff
@@ -452,6 +458,57 @@ For example, on Linux this would happen before the `pivot_root` operation is exe
 
+### <a name="configHooksSendSeccompFd" />SendSeccompFd Hooks
+
+The `sendSeccompFd` hooks MUST only be called if the seccomp policy contains `SCMP_ACT_NOTIFY`.
```
```suggestion
The `sendSeccompFd` hooks MUST only be called if the seccomp policy contains `SCMP_ACT_NOTIFY` actions.
```

mauriciovasquezbernal commented on config.md:
```diff
@@ -452,6 +458,57 @@ For example, on Linux this would happen before the `pivot_root` operation is exe
+    The value MAY be one of:
+
+    * `start`: the seccomp filter is created following the [`start`](runtime.md#start) command
+    * `exec`: the seccomp filter is created following an `exec` command
```
I just realized the OCI spec doesn't define an `exec` command. Given that, I'm not sure about including this field in the state anymore. 

Maybe we could use `State.Status` to differentiate between a call starting the container or a call when the container is already running?

alban replied:
sounds good if it works. Have you checked what does `State.Status` contain when the seccomp hook is called from the startup phase?

mauriciovasquezbernal replied:
Yes, it's "creating", hence the hook / agent is able to differentiate the two cases based on it.

alban replied:
Checked, it is State.Status=`creating` and State.Status=`running`. Looks good.

alban replied:
I removed the "phase" field.

mauriciovasquezbernal commented on specs-go/state.go:
```diff
@@ -33,3 +35,32 @@ type State struct {
+	// SeccompFd is the file descriptor for Seccomp User Notification
+	SeccompFd int `json:"seccompFd"`
+	// Pid is the process ID on which the seccomp filter is applied
+	Pid int `json:"pid,omitempty"`
```
Should it really have `omitempty` as it's required?

alban replied:
```suggestion
	Pid int `json:"pid"`
```

mauriciovasquezbernal reviewed (commented):
Some more minor comments, one important about the `exec` phase.
Should this PR also update the [Lifecycle](https://github.com/kinvolk/runtime-spec/blob/master/runtime.md#lifecycle) ?

alban committed:
fixup: Add Seccomp Notify support

Fixup following reviews

Signed-off-by: Alban Crequy <alban@kinvolk.io>

alban commented:
I updated the branch based on reviews.

> Should this PR also update the Lifecycle ?

I am not sure. I didn't update it.

mauriciovasquezbernal reviewed (approved):
LGTM!

alban commented:
Upstream PR 1073
